File: app/src/test/java/bj/discogsbrowser/artist/ArtistControllerTest.java
package bj.discogsbrowser.artist;

import android.content.Context;

import com.airbnb.epoxy.EpoxyModel;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.List;

import bj.discogsbrowser.artistreleases.ArtistReleasesFactory;
import bj.discogsbrowser.artistreleases.ArtistResultFactory;
import bj.discogsbrowser.utils.ImageViewAnimator;
import bj.discogsbrowser.utils.analytics.AnalyticsTracker;

import static android.os.Build.VERSION_CODES.LOLLIPOP;
import static junit.framework.Assert.assertEquals;

/**
 * Created by Josh Laird on 04/05/2017.
 */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = LOLLIPOP, manifest = Config.NONE)
public class ArtistControllerTest
{
    private ArtistController controller;
    @Mock Context context;
    @Mock ArtistContract.View view;
    @Mock ImageViewAnimator imageViewAnimator;
    @Mock AnalyticsTracker tracker;

    @Before
    public void setup()
    {
        MockitoAnnotations.initMocks(this);
        controller = new ArtistController(view, context, imageViewAnimator, tracker);
        controller.requestModelBuild();
    }

    @After
    public void tearDown()
    {
        // Not too fussed about the interactions, just want to make sure that the correct models build
//        verifyNoMoreInteractions(view, context, imageViewAnimator, tracker);
    }


    @Test
    public void initialLoadingState_correct()
    {
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "LoadingModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 4);
    }

    @Test
    public void setArtistNoMembers_displaysNoMembers()
    {
        controller.setArtist(ArtistReleasesFactory.getTestArtistResultNoMembers());
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "ViewReleasesModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "SubHeaderModel_");
        assertEquals(copyOfModels.get(5).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(6).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(7).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 8);
    }

    @Test
    public void setArtistTwoMembers_displaysTwoMembers()
    {
        controller.setArtist(ArtistResultFactory.buildArtistResult(2));
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "SubHeaderModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "MemberModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "MemberModel_");
        assertEquals(copyOfModels.get(5).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(6).getClass().getSimpleName(), "ViewReleasesModel_");
        assertEquals(copyOfModels.get(7).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(8).getClass().getSimpleName(), "SubHeaderModel_");
        assertEquals(copyOfModels.get(9).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(10).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(11).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 12);
    }

    @Test
    public void setArtistWithWantedUrl_displaysArtistWithWantedUrl()
    {
        controller.setArtist(ArtistReleasesFactory.getTestArtistResultWantedUrls());
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "ViewReleasesModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "SubHeaderModel_");
        assertEquals(copyOfModels.get(5).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(6).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(7).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 8);
    }

    @Test
    public void errorThenSetArtist_displaysArtist()
    {
        controller.setError(true);
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "RetryModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 4);

        controller.setLoading(true);
        copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "LoadingModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 4);

        controller.setArtist(ArtistReleasesFactory.getTestArtistResultNoMembers());
        copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "HeaderModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "ViewReleasesModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "SubHeaderModel_");
        assertEquals(copyOfModels.get(5).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(6).getClass().getSimpleName(), "UrlModel_");
        assertEquals(copyOfModels.get(7).getClass().getSimpleName(), "EmptySpaceModel_");
        assertEquals(copyOfModels.size(), 8);
    }
}

File: app/src/test/java/bj/vinylbrowser/artist/ArtistControllerTest.kt
package bj.vinylbrowser.artist

import android.content.Context
import android.os.Build.VERSION_CODES.LOLLIPOP
import bj.vinylbrowser.artistreleases.ArtistResultFactory
import bj.vinylbrowser.utils.ImageViewAnimator
import bj.vinylbrowser.utils.analytics.AnalyticsTracker
import com.nhaarman.mockito_kotlin.mock
import junit.framework.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Created by Josh Laird on 19/05/2017.
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = intArrayOf(LOLLIPOP), manifest = Config.NONE)
class ArtistControllerTest {
    private lateinit var controller: ArtistEpxController
    val context: Context = mock()
    val view: ArtistContract.View = mock()
    val imageViewAnimator: ImageViewAnimator = mock()
    val tracker: AnalyticsTracker = mock()

    @Before
    fun setUp() {
        controller = ArtistEpxController(view, context, imageViewAnimator, tracker)
        controller.requestModelBuild()
    }

    @Test
    fun initialLoadingState_correct() {
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "LoadingModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 4)
    }

    @Test
    fun setArtistNoMembers_displaysNoMembers() {
        controller.setArtist(ArtistResultFactory.buildArtistResult(0))
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "ViewReleasesModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "SubHeaderModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 8)
    }

    @Test
    fun setArtistTwoMembers_displaysTwoMembers() {
        controller.setArtist(ArtistResultFactory.buildArtistResult(2))
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "SubHeaderModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "MemberModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "MemberModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "ViewReleasesModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[8].javaClass.simpleName, "SubHeaderModel_")
        assertEquals(copyOfModels[9].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[10].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[11].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 12)
    }

    @Test
    fun setArtistWithWantedUrl_displaysArtistWithWantedUrl() {
        controller.setArtist(ArtistResultFactory.buildArtistResult(0))
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "ViewReleasesModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "SubHeaderModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 8)
    }

    @Test
    fun errorThenSetArtist_displaysArtist() {
        controller.setError(true)
        var copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "RetryModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 4)

        controller.setLoading(true)
        copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "LoadingModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 4)

        controller.setArtist(ArtistResultFactory.buildArtistResult(0))
        copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "HeaderModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "ViewReleasesModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "SubHeaderModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "UrlModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "EmptySpaceModel_")
        assertEquals(copyOfModels.size, 8)
    }
}